[docs] Add new page for query planning best practices (#7263)
apollographql/router: docs/source/routing/query-planning/query-planning-best-practices.mdx

---
title: Query Planning Best Practices
subtitle: What things should we keep in mind as we construct our supergraph and the operations that consume them?
---

> When working with Apollo Federation, changes in our schema can have unexpected impact on the complexity of our graph.

Reviewer: ```suggestion
When working with Apollo Federation, changes in your schema can have unexpected impact on the complexity of our graph.
```

Complexity how? Query plans?

> While many issues can be found at build time with [schema composition](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/composition), others changes may lead to issues that only arise at runtime while generating or executing query plans.

Reviewer: I believe links can be relative here, and don't need the full domain, although it's not going to harm anything here.

Reviewer: Also, it may be worth expounding on what issues may arise- to my knowledge, composition should be catching them all and us flagging that QP may have issues is not possibly correct here.

Examples of changes that can impact query planning include:

> * Modifying `@key`, `@requires`, `@provides`, or `@shareable` directive usage

Reviewer: ```suggestion
* Adding or modifying `@key`, `@requires`, `@provides`, or `@shareable` directive usage
```

* Adding or removing a type implementation from an interface
* Using `interfaceObject` and adding new fields to an interface

To help alleviate these issues as much as possible, we recommend following some of these best practices for your federated graph.

## Shared Types and Fields

The [`@shareable` directive](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/sharing-types) allows multiple subgraphs to resolve the same types or fields on entities, giving the query planner options for potentially shorter query paths. However, it's important to use it judiciously.

> - Extensive use can exponentially increase the number of possible query plans generated, potentially leading to performance degradation at runtime as we generate uncached plans.

Reviewer: ```suggestion
- Extensive `@shareable` use can exponentially increase the number of possible query plans generated as the query planner will find the shortest path to the desired result. This can then potentially lead to performance degradation at runtime as we generate plans.
```

> - Using `@shareable` at root fields on the `Query`, `Mutation`, and `Subscription` types indicates that any subgraph can resolve a given entry point. While query plans can be deterministic for a given version of Router + Federation, there are no guarantees across versions, meaning that your plans may change if new services get added or deleted. This could cause an unexpected change in traffic a service even there were no changes in the operations.

Reviewer: It's worth caveating that if a root query returns a list, the router may (and is likely to) run a query in parallel and expects the results to be in the _exact same order_.

## Operations Spanning Multiple Subgraphs

Operations that need to query multiple subgraphs can impact performance because each additional subgraph queried adds complexity to the query plan, increasing the time in the Router for both generation and execution of the operation.
- Design your schema to minimize operations that span numerous subgraphs.
- Using directives like `@requires` or `@interfaceObject` carefully to control complexity.

> ## `@requires` Directive

Reviewer: Is this supposed to be ### or just ##?

Author: I had use H2 as the section for all the things to watch our for and recommended steps to monitor and fix

The [`@requires` directive](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/reference/directives#requires) allows a subgraph to fetch additional fields needed to resolve an entity. This can be powerful but must be handled with care.

> - Changes to fields utilized by `@requires` can impact the subgraph fetches that current operations depend on and may create larger and slower plans

Reviewer: ```suggestion
- Changes to fields utilized by `@requires` can impact the subgraph fetches that current operations depend on and may create larger and slower plans.
```

2. Deploy a new version of the `Reviews` subgraph with a resolver that accepts either `nameLowercase` or `name` in the source object.
3. Modify the Reviews subgraph's schema in the registry so that it `@requires(fields: "name")`.
4. Deploy a new version of the `Reviews` subgraph with a resolver that only accepts the `name` in its source object.

Alternatively, you can perform this operation with an atomic migration at the subgraph level by modifying the subgraph's URL:

1. Modify the `Products` subgraph to add the `name` field (as usual, first deploy all replicas, then use `rover subgraph publish` to push the new subgraph schema).
2. Deploy a new set of `Reviews` replicas to a new URL that reads from `name`.
3. Register the `Reviews` subgraph with the new URL and the schema changes above.

With this atomic strategy, the query planner resolves all outstanding requests to the old subgraph URL that relied on `nameLowercase` with the old query-planning configuration, which `@requires` the `nameLowercase` field. All new requests are made to the new subgraph URL using the new query-planning configuration, which `@requires` the `name` field.

> ## Interface Implementations

Reviewer: Same note about headings as above.

---

> title: Query Planning Best Practices

Reviewer 2: ```suggestion
title: Best Practices for Query Planning
```

> subtitle: What things should we keep in mind as we construct our supergraph and the operations that consume them?

Reviewer 2: ```suggestion
subtitle: Design your schemas to optimize query planning performance
description: Learn best practices in GraphQL schema design to achieve efficient query planning of your graphs using Apollo Federation and Apollo Router
```

---

> When working with Apollo Federation, changes in your schema can have unexpected impact on the complexity of our graph. Adding one field or changing one directive may create a new supergraph that has hundreds or even thousands of new possible paths and edges to connect entities and resolve client operations.

Reviewer 2: Try adding impacts on performance

```suggestion
When working with Apollo Federation, changes in your schema can have unexpected impact on the complexity and performance of your graph. Adding one field or changing one directive may create a new supergraph that has hundreds, or even thousands, of new possible paths and edges to connect entities and resolve client operations. Consequently, query planning throughput and latency may degrade.
```

> While validation errors can be found at build time with [schema composition](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/composition), others changes may lead to issues that only arise at runtime while generating or executing query plans.

Reviewer 2: ```suggestion
While validation errors can be found at build time with [schema composition](https://www.apollographql.com/docs/graphos/schema-design/federated-schemas/composition), other changes may lead to issues that only arise at runtime, during query plan generation or execution.
```

Examples of changes that can impact query planning include:
* Adding or modifying `@key`, `@requires`, `@provides`, or `@shareable` directive usage
* Adding or removing a type implementation from an interface
* Using `interfaceObject` and adding new fields to an interface

To help alleviate these issues as much as possible, we recommend following some of these best practices for your federated graph.

> ## Shared Types and Fields

Reviewer 2: Try making a best-practice recommendation in each heading, it'll help scannability and SEO

```suggestion
## Use shared types and fields judiciously
```

- Extensive `@shareable` use can exponentially increase the number of possible query plans generated as the query planner will find the shortest path to the desired result. This can then potentially lead to performance degradation at runtime as we generate plans.
- Using `@shareable` at root fields on the `Query`, `Mutation`, and `Subscription` types indicates that any subgraph can resolve a given entry point. While query plans can be deterministic for a given version of Router + Federation, there are no guarantees across versions, meaning that your plans may change if new services get added or deleted. This could cause an unexpected change in traffic for a given service, even there were no changes in the operations.
  - Using shared root types also implies that the fields return the same data in the same order across all subgraphs, even if the data is a list, which is often not the case for dynamic applications.

> ## Operations Spanning Multiple Subgraphs

Reviewer 2: ```suggestion
## Minimize operations spanning multiple subgraphs
```

2. Deploy a new version of the `Reviews` subgraph with a resolver that accepts either `nameLowercase` or `name` in the source object.
3. Modify the Reviews subgraph's schema in the registry so that it `@requires(fields: "name")`.
4. Deploy a new version of the `Reviews` subgraph with a resolver that only accepts the `name` in its source object.

Alternatively, you can perform this operation with an atomic migration at the subgraph level by modifying the subgraph's URL:

1. Modify the `Products` subgraph to add the `name` field (as usual, first deploy all replicas, then use `rover subgraph publish` to push the new subgraph schema).
2. Deploy a new set of `Reviews` replicas to a new URL that reads from `name`.
3. Register the `Reviews` subgraph with the new URL and the schema changes above.

With this atomic strategy, the query planner resolves all outstanding requests to the old subgraph URL that relied on `nameLowercase` with the old query-planning configuration, which `@requires` the `nameLowercase` field. All new requests are made to the new subgraph URL using the new query-planning configuration, which `@requires` the `name` field.

> ### Interface Implementations

Reviewer 2: Should this be an H2 heading (and it's subsection also one level up)?

```suggestion
## Manage interface migrations
```

Interfaces are an essential part of GraphQL schema design, offering flexibility in defining polymorphic types. However, they can also be open for implementation across service boundaries, allowing subgraphs to contribute a new type that changes how existing operations execute.

- Approach interface migrations similar to database migrations. Ensure that changes to interface implementations are performed safely, avoiding disruptions to query operations.

> #### Example

Reviewer 2: ```suggestion
### Example
```

2. Deploy a new version of the subgraph that removes the `EmailChannel` type.

The first step causes the query planner to stop sending fragments `...on EmailChannel`, which would fail validation if sent to a subgraph that isn't aware of the type.

If you want to keep the `EmailChannel` type but remove it from the `Channel` interface, the process is similar. Instead of removing the `EmailChannel` type altogether, only remove the `implements Channel` addendum to the type definition. This is because the query planner expands queries to interfaces or unions into fragments on their implementing types.

For example, a query like this:

```graphql
query FindChannel($id: ID!) {
  channel(id: $id) {
    id
  }
}
```

generates two queries, one to each subgraph, like so:

<CodeColumns>

    ```graphql title="Query to email subgraph"
    query {
    _entities(...) {
    ...on EmailChannel {
    id
}
}
}
    ```

    ```graphql title="Query to web subgraph"
    query {
    _entities(...) {
    ...on WebChannel {
    id
}
}
}
    ```

</CodeColumns>

Currently, the router expands all interfaces into implementing types.

> ## Recommended Steps

Reviewer 2: ```suggestion
## Use recommended features

GraphOS and router provide many features that help monitor and improve query planning performance, both at build time and runtime. 
```

### Runtime
In the [Router configuration](https://www.apollographql.com/docs/graphos/routing/configuration) there are many settings to help monitor and improve performance impacts. Here are some features all production graphs should consider:

* Monitor your query planner performance with the [standard instruments](https://www.apollographql.com/docs/graphos/routing/observability/telemetry/instrumentation/standard-instruments#query-planning)
* Enabling and configuring the [in-memory cache for query plans](https://www.apollographql.com/docs/graphos/routing/performance/caching/in-memory)
* Using the cache [warm up features](https://www.apollographql.com/docs/graphos/routing/performance/caching/in-memory#cache-warm-up) included out of the box and using the `dry-run` headers for operations
* Enabling and configuring [distributed caches for query plans](https://www.apollographql.com/docs/graphos/routing/performance/caching/distributed) to share across router instances

> * Limiting the size of operations (and therefore their query plans) with [request limits](https://www.apollographql.com/docs/graphos/routing/security/request-limits) and the cost with [demand control](https://www.apollographql.com/docs/graphos/routing/security/demand-control)

Reviewer 2: Consider putting Runtime section after Build time?

> ### Build Time

Reviewer 2: ```suggestion
### Build time
```

- Using `@shareable` at root fields on the `Query`, `Mutation`, and `Subscription` types indicates that any subgraph can resolve a given entry point. While query plans can be deterministic for a given version of Router + Federation, there are no guarantees across versions, meaning that your plans may change if new services get added or deleted. This could cause an unexpected change in traffic for a given service, even there were no changes in the operations.
  - Using shared root types also implies that the fields return the same data in the same order across all subgraphs, even if the data is a list, which is often not the case for dynamic applications.

## Minimize operations spanning multiple subgraphs

Operations that need to query multiple subgraphs can impact performance because each additional subgraph queried adds complexity to the query plan, increasing the time in the Router for both generation and execution of the operation.
- Design your schema to minimize operations that span numerous subgraphs.
- Using directives like `@requires` or `@interfaceObject` carefully to control complexity.

> ### `@requires` Directive

Reviewer 2: ```suggestion
### `@requires` directive
```

  channel(id: $id) {
    id
  }
}
```

generates two queries, one to each subgraph, like so:

<CodeColumns>

    ```graphql title="Query to email subgraph"
    query {
    _entities(...) {
    ...on EmailChannel {
    id
}
}
}
    ```

    ```graphql title="Query to web subgraph"
    query {
    _entities(...) {
    ...on WebChannel {
    id
}
}
}
    ```

</CodeColumns>

Currently, the router expands all interfaces into implementing types.

## Use recommended features

GraphOS and router provide many features that help monitor and improve query planning performance, both at build time and runtime. 

### Build time

* Use [schema proposals](https://www.apollographql.com/docs/graphos/platform/schema-management/proposals) to review changes that have a large impact across entities and interfaces
* Enable [common linter settings](https://www.apollographql.com/docs/graphos/platform/schema-management/linting)
* Setup [custom checks](https://www.apollographql.com/docs/graphos/platform/schema-management/checks/custom) to do advanced and specific validations, like [limiting the size of query plans](https://github.com/apollosolutions/example-graphos-custom-check-query-planner)

### Runtime

> In the [Router configuration](https://www.apollographql.com/docs/graphos/routing/configuration) there are many settings to help monitor and improve performance impacts. Here are some features all production graphs should consider:

Reviewer 2: ```suggestion
In the [router configuration](https://www.apollographql.com/docs/graphos/routing/configuration) there are many settings to help monitor and improve performance impacts. Here are some features all production graphs should consider:
```
